File: src/core/ThemeDirector.ts
import type Phaser from 'phaser';
import { GameEvent, type GameSystem } from './contracts';
import type { EventBus } from './EventBus';
import { paletteNameForLevel } from './Progression';
import { PALETTES, Theme, applyPalette, type PaletteName } from './Theme';
import { lerpPalette } from './themeMath';

/** Cross-fade duration — matches ArenaView's ZOOM_MS so the palette lands with the zoom. */
const FADE_MS = 1200;

/**
 * Scene-level owner of the milestone colour swap (shared shell, like the PhaseDirector).
 *
 * PROGRESSION_CHANGED tells it which authored palette the run has earned
 * (`paletteNameForLevel`); the next ARENA_ZOOM {active:true} — only ever emitted by the
 * milestone zoom, whose input freeze we piggyback — cross-fades the active `Theme` there,
 * emitting THEME_CHANGED per tick so every baked surface restyles in step with the zoom.
 * It never touches a zone: `Theme` is the shared data, the bus is the repaint signal.
 */
export class ThemeDirector implements GameSystem {
  private scene?: Phaser.Scene;
  private current: PaletteName = 'workshop';
  private target: PaletteName = 'workshop';
  private tween?: Phaser.Tweens.Tween;

  constructor(private readonly bus: EventBus) {}

  create(scene: Phaser.Scene): void {
    this.scene = scene;
    // GameScene resets Theme to workshop before anything bakes a colour; mirror that truth.
    this.current = 'workshop';
    this.target = 'workshop';

    this.bus.on(GameEvent.ProgressionChanged, ({ level }) => {
      this.target = paletteNameForLevel(level);
    });
    this.bus.on(GameEvent.ArenaZoom, ({ active }) => {
      if (active) this.beginFade();
    });
  }

  update(_time: number, _delta: number): void {}

  destroy(): void {
    this.tween?.remove();
    this.tween = undefined;
  }

  private beginFade(): void {
    if (this.target === this.current) return;
    const from = { ...Theme };
    const to = PALETTES[this.target];
    this.current = this.target;

    this.tween?.remove();
    const proxy = { t: 0 };
    this.tween = this.scene?.tweens?.add({
      targets: proxy,
      t: 1,
      duration: FADE_MS,
      ease: 'Sine.easeInOut',
      onUpdate: () => {
        applyPalette(lerpPalette(from, to, proxy.t));
        this.bus.emit(GameEvent.ThemeChanged);
      },
      onComplete: () => {
        this.tween = undefined;
        applyPalette(to);
        this.bus.emit(GameEvent.ThemeChanged);
      },
    });
    // Headless/edge case: no tweens available — snap so the palette can't be left stale.
    if (!this.tween) {
      applyPalette(to);
      this.bus.emit(GameEvent.ThemeChanged);
    }
  }
}

File: Assets/Game/Gameplay/ThemeDirector.cs
using PrimeTween;
using RichCoast.Core;

namespace RichCoast.Game
{
    /// <summary>
    /// Scene-level owner of the milestone colour swap (port of <c>ThemeDirector.ts</c>).
    /// <c>ProgressionChanged</c> tells it which authored palette the run has earned
    /// (<see cref="ProgressionCurve.PaletteNameForLevel"/>); the next <c>ArenaZoom(true)</c> — only
    /// ever raised by the milestone zoom, whose input freeze we piggyback — cross-fades the active
    /// <see cref="Theme"/> there, raising <c>ThemeChanged</c> per tick so every bound surface
    /// restyles in step with the zoom. It never touches a zone: <see cref="Theme"/> is the shared
    /// data, the event is the repaint signal.
    /// </summary>
    public sealed class ThemeDirector
    {
        readonly ProgressionCurve curve;
        readonly GameFeelSO feel;
        string current = Palettes.WorkshopName;
        string target = Palettes.WorkshopName;
        Tween tween;

        public string CurrentPaletteName => current;

        public ThemeDirector(ProgressionCurve curve, GameFeelSO feel)
        {
            this.curve = curve;
            this.feel = feel;
            GameEvents.ProgressionChanged += e => target = curve.PaletteNameForLevel(e.Level);
            GameEvents.ArenaZoom += active => { if (active) BeginFade(); };
        }

        /// <summary>
        /// Jump straight to a level's palette with no cross-fade — a restored run boots mid-progression.
        /// Without this the board would stay painted workshop until the next milestone zoom, because
        /// <see cref="BeginFade"/> is the only thing that ever moves <see cref="current"/>.
        /// Call AFTER the progression has been announced: that announcement is what sets <c>target</c>.
        /// </summary>
        public void SnapTo(int level)
        {
            tween.Stop();
            current = target = curve.PaletteNameForLevel(level);
            Theme.Apply(Palettes.Get(current));
            GameEvents.RaiseThemeChanged();
        }

        void BeginFade()
        {
            if (target == current) return;
            var from = Theme.Active;
            var to = Palettes.Get(target);
            current = target;

            tween.Stop();
            float seconds = feel.milestoneZoomMs / 1000f;
            if (seconds <= 0f)
            {
                Theme.Apply(to);
                GameEvents.RaiseThemeChanged();
                return;
            }
            tween = Tween.Custom(this, 0f, 1f, seconds, (self, t) =>
            {
                Theme.Apply(Palette.Lerp(from, to, t));
                GameEvents.RaiseThemeChanged();
            }, Ease.InOutSine).OnComplete(this, self =>
            {
                Theme.Apply(to);
                GameEvents.RaiseThemeChanged();
            }, warnIfTargetDestroyed: false);
        }
    }
}
